### PR TITLE
Create tritium-tags module

### PR DESCRIPTION
**Pull request by schlosna**
Strawman example tagging API

TODO:

- [ ] finalize format & parser
- [ ] `@Safe` & `@Unsafe` tags
- [ ] wire through `Instrumentation` and maybe `Tritium`

### CODE REVIEW

File: tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+        get = {"get*", "is*"},
+        // Try to avoid leaking Immutable prefixed objects to avoid
+        // api fallout when we switch to another processor.
+        // Side effect: Return the base type from builders.
+        // Side effect: Nest implementation class inside of builder class.
+        visibility = Value.Style.ImplementationVisibility.PRIVATE,
+        builderVisibility = Value.Style.BuilderVisibility.PACKAGE,
+        // Default interface methods don't need to be annotated @Value.Default
+        defaultAsDefault = true)
+public abstract class TaggedMetric {
```
Comment (ellisjoe) [inline tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java:44]: should this just be a static util class? similar to `MetricRegistry#name`?

**Comment by ellisjoe** [comment]
think you can avoid wiring this through if we make this a static util and add a listener when we create metric registries that validates names like we talked about

File: tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+     */
+    @Value.Lazy
+    public String canonicalName() {
+        if (nonEmptyTags().isEmpty()) {
+            return name();
+        } else {
+            StringBuilder builder = new StringBuilder();
+            builder.append(name()).append('[');
+
+            for (Iterator<Map.Entry<String, String>> it = nonEmptyTags().entrySet().iterator(); it.hasNext(); ) {
```
Comment (j-baker) [inline tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java]: I think you'll benefit at some point from sorting the keys (e.g. if you make a map of these things, you'll be able to lookup).

Comment (j-baker) [reply tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java]: like presumably the user is going to be like metricRegistry.meter(TaggedMetric.of("name", singletonMap("tag", "value"))

**Comment by ellisjoe** [comment]
looks good to me

File: tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java (hunk 1 above)
Comment (schlosna) [reply tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java:44]: `TaggedMetric.toCanonicalName(String name, Map<String, String> tags)` static method will now generate the canonical string name for given metric name and tags.

File: tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java (hunk 2 above)
Comment (schlosna) [reply tritium-tags/src/main/java/com/palantir/tritium/tags/TaggedMetric.java]: applied sorting and some normalization of keys (lower casing, require <= 20 alpha numeric)

**Pull request merged**